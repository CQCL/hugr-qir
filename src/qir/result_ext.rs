use anyhow::{bail, Result, anyhow};
use hugr::{extension::{prelude::ConstString, simple_op::MakeExtensionOp as _}, ops::ExtensionOp, HugrView};
use hugr_llvm::{emit::{emit_value, EmitFuncContext, EmitOpArgs}, inkwell::types::BasicType as _, sum::LLVMSumValue, types::HugrSumType};
use tket2_hseries::extension::result::{ResultOp, ResultOpDef};

use super::QirCodegenExtension;
impl QirCodegenExtension {
    pub fn emit_result_op<'c,H: HugrView>(&self, context: &mut EmitFuncContext<'c,'_,H>, args: EmitOpArgs<'c,'_,ExtensionOp,H>, op: ResultOpDef) -> Result<()> {
        let result_op = ResultOp::from_extension_op(&args.node())?;
        let tag_str = result_op.tag;
        if tag_str.is_empty() {
            bail!("Empty result tag received")
        }

        let tag_ptr = emit_value(context, &ConstString::new(tag_str).into())?;
        let i8_ptr_ty = context
            .iw_context()
            .i8_type()
            .ptr_type(Default::default())
            .as_basic_type_enum();

        match op {
            ResultOpDef::Bool => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let bool_type = context.llvm_sum_type(HugrSumType::new_unary(2))?;
                let val = LLVMSumValue::try_new(val, bool_type)
                    .map_err(|_| anyhow!("bool_type expects a value"))?
                    .build_get_tag(context.builder())?;
                let i1_ty = context.iw_context().bool_type();
                let trunc_val = context.builder().build_int_truncate(val, i1_ty, "")?;
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i1_ty.into(), i8_ptr_ty.into()], false);
                let print_fn = context.get_extern_func(
                    "__quantum__rt__bool_record_output",
                    print_fn_ty,
                )?;
                context.builder().build_call(
                    print_fn,
                    &[trunc_val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::Int | ResultOpDef::UInt => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let i64_ty = context.iw_context().i64_type();
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn = context
                    .get_extern_func("__quantum__rt__int_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::F64 => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let f64_ty = context.iw_context().f64_type();
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[f64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn = context.get_extern_func(
                    "__quantum__rt__double_record_output",
                    print_fn_ty,
                )?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::ArrBool => todo!(),
            ResultOpDef::ArrInt => todo!(),
            ResultOpDef::ArrUInt => todo!(),
            ResultOpDef::ArrF64 => todo!(),
            _ => todo!(),
        }

    }

}
            // .simple_extension_op::<tket2_hseries::extension::result::ResultOpDef>(
            //     |context, args, op| {
            //         let result_op = ResultOp::from_extension_op(&args.node())?;
            //         let tag_str = result_op.tag;
            //         if tag_str.is_empty() {
            //             return Err(anyhow!("Empty result tag received"));
            //         }

            //         let tag_ptr = emit_value(context, &ConstString::new(tag_str).into())?;
            //         let i8_ptr_ty = context
            //             .iw_context()
            //             .i8_type()
            //             .ptr_type(Default::default())
            //             .as_basic_type_enum();

            //         match op {
            //             ResultOpDef::Bool => {
            //                 let [val] = args
            //                     .inputs
            //                     .try_into()
            //                     .map_err(|_| anyhow!("result_bool expects one input"))?;
            //                 let bool_type = context.llvm_sum_type(HugrSumType::new_unary(2))?;
            //                 let val = LLVMSumValue::try_new(val, bool_type)
            //                     .map_err(|_| anyhow!("bool_type expects a value"))?
            //                     .build_get_tag(context.builder())?;
            //                 let i1_ty = context.iw_context().bool_type();
            //                 let trunc_val = context.builder().build_int_truncate(val, i1_ty, "")?;
            //                 let print_fn_ty = context
            //                     .iw_context()
            //                     .void_type()
            //                     .fn_type(&[i1_ty.into(), i8_ptr_ty.into()], false);
            //                 let print_fn = context.get_extern_func(
            //                     "__quantum__rt__bool_record_output",
            //                     print_fn_ty,
            //                 )?;
            //                 context.builder().build_call(
            //                     print_fn,
            //                     &[trunc_val.into(), tag_ptr.into()],
            //                     "print_bool",
            //                 )?;
            //                 args.outputs.finish(context.builder(), [])
            //             }
            //             ResultOpDef::Int | ResultOpDef::UInt => {
            //                 let [val] = args
            //                     .inputs
            //                     .try_into()
            //                     .map_err(|_| anyhow!("result_bool expects one input"))?;
            //                 let i64_ty = context.iw_context().i64_type();
            //                 let print_fn_ty = context
            //                     .iw_context()
            //                     .void_type()
            //                     .fn_type(&[i64_ty.into(), i8_ptr_ty.into()], false);
            //                 let print_fn = context
            //                     .get_extern_func("__quantum__rt__int_record_output", print_fn_ty)?;
            //                 context.builder().build_call(
            //                     print_fn,
            //                     &[val.into(), tag_ptr.into()],
            //                     "print_bool",
            //                 )?;
            //                 args.outputs.finish(context.builder(), [])
            //             }
            //             ResultOpDef::F64 => {
            //                 let [val] = args
            //                     .inputs
            //                     .try_into()
            //                     .map_err(|_| anyhow!("result_bool expects one input"))?;
            //                 let f64_ty = context.iw_context().f64_type();
            //                 let print_fn_ty = context
            //                     .iw_context()
            //                     .void_type()
            //                     .fn_type(&[f64_ty.into(), i8_ptr_ty.into()], false);
            //                 let print_fn = context.get_extern_func(
            //                     "__quantum__rt__double_record_output",
            //                     print_fn_ty,
            //                 )?;
            //                 context.builder().build_call(
            //                     print_fn,
            //                     &[val.into(), tag_ptr.into()],
            //                     "print_bool",
            //                 )?;
            //                 args.outputs.finish(context.builder(), [])
            //             }
            //             ResultOpDef::ArrBool => todo!(),
            //             ResultOpDef::ArrInt => todo!(),
            //             ResultOpDef::ArrUInt => todo!(),
            //             ResultOpDef::ArrF64 => todo!(),
            //             _ => todo!(),
            //         }
            //     },
            // )
