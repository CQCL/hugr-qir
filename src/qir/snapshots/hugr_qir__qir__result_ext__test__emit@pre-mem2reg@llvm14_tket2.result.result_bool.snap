---
source: src/qir/result_ext.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

@0 = private unnamed_addr constant [9 x i8] c"bool_int\00", align 1

define void @_hl.main.1({ i32, {}, {} } %0) {
alloca_block:
  %"2_0" = alloca { i32, {}, {} }, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store { i32, {}, {} } %0, { i32, {}, {} }* %"2_0", align 4
  %"2_01" = load { i32, {}, {} }, { i32, {}, {} }* %"2_0", align 4
  %1 = extractvalue { i32, {}, {} } %"2_01", 0
  %2 = trunc i32 %1 to i1
  call void @__quantum__rt__bool_record_output(i1 %2, i8* getelementptr inbounds ([9 x i8], [9 x i8]* @0, i32 0, i32 0))
  ret void
}

declare void @__quantum__rt__bool_record_output(i1, i8*)
